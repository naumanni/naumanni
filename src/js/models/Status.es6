import moment from 'moment'
import {is, fromJS, List, Map, Record} from 'immutable'

import {
  MESSAGE_TAG_REX,
  VISIBLITY_DIRECT, VISIBLITY_PRIVATE, VISIBLITY_UNLISTED, VISIBLITY_PUBLIC,
} from 'src/constants'
import {compareDateForTL, emojify, parseMastodonHtml, parsedHtmlToText} from 'src/utils'
import Attachment from './Attachment'


const TagRecord = Record({  // eslint-disable-line new-cap
  name: '',
  // url: new Map(),    // hostによって値が違うのでomitする
})

const MentionRecord = Record({  // eslint-disable-line new-cap
  url: '',
  username: '',
  acct: '',
  // id: 0,   // hostによって値が違うのでomitする
})

const ApplicationRecord = Record({  // eslint-disable-line new-cap
  name: '',
  website: '',
})

const StatusRecord = Record({  // eslint-disable-line new-cap
  id_by_host: new Map(),
  uri: '',
  url: '',
  content: '',
  created_at: '',
  account: '',
  reblogs_count: '',
  favourites_count: '',
  sensitive: '',
  spoiler_text: '',
  visibility: '',
  media_attachments: new List(),
  mentions: new List(),
  tags: new List(),
  application: new ApplicationRecord(),
  reblog: null,
  in_reply_to_id_by_host: new Map(),
  in_reply_to_account_id_by_host: new Map(),
  reblogged_by_acct: new Map(),
  favourited_by_acct: new Map(),

  // naumanni
  fetched_at: null,  // WebSocketから取得した日付
  extended: new Map(),  // pluginが追加する値
})

const NOT_SET = {}


/**
 * MastodonのStatus
 */
export default class Status extends StatusRecord {
  /**
   * @constructor
   * @param {object} raw
   */
  constructor(raw, {isOriginal}={}) {
    raw = {
      ...raw,
      id_by_host: new Map(raw.id_by_host),
      in_reply_to_id_by_host: new Map(raw.in_reply_to_id_by_host),
      in_reply_to_account_id_by_host: new Map(raw.in_reply_to_account_id_by_host),
      reblogged_by_acct: new Map(raw.reblogged_by_acct),
      favourited_by_acct: new Map(raw.favourited_by_acct),
      sensitive: !!raw.sensitive,
      media_attachments: new List((raw.media_attachments || []).map((obj) => new Attachment(obj))),
      tags: new List((raw.tags || []).map((obj) => new TagRecord(obj))),
      mentions: new List((raw.mentions || []).map((obj) => new MentionRecord(obj))),
      application: new ApplicationRecord(raw.application),  // TODO: Recordにする
      extended: fromJS(raw.extended || {}),
    }

    super(raw)
    this.isOriginal = isOriginal || false
  }

  // とりあえず
  get hosts() {
    return this.id_by_host.keySeq().toArray()
  }

  get id() {
    console.error('deprecated attribute')
    require('assert')(0)
  }

  getIdByHost(host) {
    return this.id_by_host.get(host)
  }

  getInReplyToIdByHost(host) {
    return this.in_reply_to_id_by_host.get(host)
  }

  get parsedContent() {
    if(!this._parsedContent) {
      const mentions = this.mentions
      const content = emojify(this.content)
      this._parsedContent = new List(parseMastodonHtml(content, mentions))
    }
    return this._parsedContent
  }

  get plainContent() {
    return parsedHtmlToText(this.parsedContent)
  }

  get createdAt() {
    if(!this._createdAt)
      this._createdAt = moment(this.created_at)
    return this._createdAt
  }

  get fetchedAt() {
    if(!this._fetchedAt)
      this._fetchedAt = this.fetched_at ? moment(this.fetched_at) : null
    return this._fetchedAt
  }

  get hasSpoilerText() {
    return this.spoiler_text.length > 0
  }

  get spoilerText() {
    return this.spoiler_text
  }

  canReblog() {
    return (this.visibility === VISIBLITY_PUBLIC || this.visibility === VISIBLITY_UNLISTED)
      ? true
      : false
  }

  isRebloggedAt(acct) {
    return this.reblogged_by_acct.get(acct)
  }

  isFavouritedAt(acct) {
    return this.favourited_by_acct.get(acct)
  }

  /**
   * そいつあてのMentionが含まれているか？
   * @param {URI} uri そいつ
   * @return {bool}
   */
  isMentionToURI(uri) {
    if(this.mentions.find((m) => m.url === uri))
      return true
    return false
  }

  checkMerge(newObj) {
    if(is(this, newObj)) {
      return {isChanged: false, merged: this}
    }

    // mergeする。originalの方が優先。どっちも??であれば、next
    const merged = this.withMutations((self) => {
      newObj.forEach((next, key) => {
        self.update(key, NOT_SET, (prev) => {
          let result = next

          if(key === 'fetched_at') {
            // fetched_atは一番古い日付を使う. nullだったら、WebSocket以前にもってたってことなのでnullのまま
            if(!next || (prev && next && prev < next))
              result = prev
          } else if(prev instanceof Map) {
            // mapだったらnext優先でmergeする
            result = prev.mergeDeep(next)
          } else {
            // それ以外はoriginal優先
            if(this.isOriginal)
              result = prev
            else if(newObj.isOriginal)
              result = next
          }

          return result
        })
      })
    })
    merged.isOriginal = this.isOriginal || newObj.isOriginal

    return {isChanged: true, merged}
  }

  static compareForTimeline(a, b) {
    const af = a.fetchedAt
    const bf = b.fetchedAt

    if(af && bf)
      return compareDateForTL(af, bf)
    else if(!af && bf)
      return 1
    else if(af && !bf)
      return -1
    else
      return compareDateForTL(a.createdAt, b.createdAt)
  }

  // naumanni用機能
  get messageBlockInfo() {
    const match = this.content.match(MESSAGE_TAG_REX)
    if(!match)
      return null

    return {
      checksum: match[1],
      index: +match[2],
      total: +match[3],
    }
  }

  getExtended(key) {
    return this.extended.get(key)
  }
}
