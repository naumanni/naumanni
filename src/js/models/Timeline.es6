import {List} from 'immutable'
import {
  TIMELINE_FILTER_BOOST, TIMELINE_FILTER_REPLY,
} from 'src/constants'
import Notification from './Notification'
import Status from './Status'
import ChangeEventEmitter from 'src/utils/EventEmitter'


/**
 * Timelineです。
 * TODO: Modelにうつす
 */
class Timeline extends ChangeEventEmitter {
  constructor(max) {
    super()

    this._max = max
    // this._timeline = new List()
    // null means non loaded timeline, empty List means empty timeline
    this._originalTimeline = this._timeline = null
    this._filters = new Map()
  }

  /**
   * StatusをTimelineに追加する
   * @private
   * @param {List<Ref>} newRefs
   * @param {object} options
   * @return {List<Ref>} mergeされなかったrefs
   */
  push(newRefs, options) {
    let result

    // lockされてないので、primary timelineにmergeする
    const {merged, removes} = mergeTimeline(this.timeline, newRefs, this.compare, this._max)

    result = removes
    if(!merged.equals(this._timeline)) {
      this._update(merged)
    }

    return result
  }

  /**
   * Statusフィルタをupdate
   * @param {Map<string, boolean>} filters
   */
  updateFilter(filters) {
    this._filters = filters
    this._originalTimeline && this._update(this._originalTimeline)
  }

  _update(newRefs) {
    if(!newRefs.equals(this._originalTimeline)) {
      this._originalTimeline = newRefs
    }
    this._timeline = this._filtered(newRefs)
    this.emitChange()
  }

  /**
   * Statusをフィルタする
   * @param {List<Ref>} refs
   * @return {List<Ref>} フィルタされたrefs
   */
  _filtered(refs) {
    for(const [type, toggle] of this._filters.entries()) {
      if(toggle)
        continue

      if(type === TIMELINE_FILTER_BOOST) {
        refs = refs.filter((ref) => (
          ![...ref.resolve().reblogged_by_acct.values()]
            .includes(true))
        )
      } else if(type === TIMELINE_FILTER_REPLY) {
        refs = refs.filter((ref) => (
          ![...ref.resolve().in_reply_to_account_id_by_host.values()]
            .reduce((prev, id) => prev && id != null, true)
        ))
      }
    }

    return refs
  }

  get timeline() {
    return this._timeline || new List()
  }

  // timelineの中に含まれるuriのリストを返す
  get uris() {
    require('assert')(0, 'not implemented')
  }

  getMinIdForHost(host) {
    let minId = null

    for(const ref of this._timeline) {
      const id = ref.resolve().getIdByHost(host)
      if(id && (!minId || id < minId)) {
        minId = id
      }
    }

    return minId
  }

  /**
   * 自分のCloneを返す。EventListenerはCloneしない
   * @return {Timeline}
   */
  clone() {
    const newTL = new this.constructor(this._max)

    newTL._originalTimeline = newTL._timeline = this._timeline
    return newTL
  }

  set max(newMax) {
    this._max = newMax
  }

  hasChanges({accounts, statuses}) {
    return true
  }

  // private
  compare() {
    require('assert')('not implemented')
  }
}


export class StatusTimeline extends Timeline {
  get uris() {
    return this.timeline.map((ref) => ref.uri)
  }

  hasChanges({accounts, statuses}) {
    return this.timeline.find((ref) => {
      if(statuses[ref.uri] || accounts[ref.accountUri])
        return true
      const {reblogRef} = ref
      if(reblogRef && (statuses[reblogRef.uri] || accounts[reblogRef.accountUri]))
        return true
      return false
    }) ? true : false
  }

  compare(a, b) {
    return Status.compareForTimeline(a.resolve(), b.resolve())
  }
}


export class NotificationTimeline extends Timeline {
  get uris() {
    const self = this
    return Array.from((function* () {
      for(const ref of self.timeline) {
        const accountRef = ref.accountRef
        if(accountRef)
          yield accountRef.uri
        const statusRef = ref.statusRef
        if(statusRef)
          yield statusRef.uri
      }
    })())
  }

  hasChanges({accounts, statuses}) {
    return this.timeline.find((ref) => {
      const accountRef = ref.accountRef
      if(accountRef && (accounts[accountRef.uri]))
        return true
      const statusRef = ref.statusRef
      if(statusRef && (statuses[statusRef.uri] || accounts[statusRef.accountUri]))
        return true
      const reblogRef = statusRef && statusRef.reblogRef
      if(reblogRef && (statuses[reblogRef.uri] || accounts[reblogRef.accountUri]))
        return true
      return false
    }) ? true : false
  }

  compare(a, b) {
    return Notification.compareForTimeline(a, b)
  }
}


/**
 * iterをtesterの返り値で２つに分ける
 * @param {Iterable} iter
 * @param {func} tester
 * @return {List} t true
 * @return {List} f false
 */
function filterSplit(iter, tester) {
  let t = new List()
  let f = new List()

  t = t.asMutable()
  f = f.asMutable()
  for(const value of iter) {
    (tester(value) ? t : f).push(value)
  }
  t = t.asImmutable()
  f = f.asImmutable()
  return {t, f}
}


function mergeTimeline(list, iter, comparator, max) {
  let {t: removes, f: appends} = filterSplit(iter, (ref) => list.find((a) => a.uri === ref.uri) ? true : false)

  if(appends.isEmpty())
    return {merged: list, appends, removes}

  let merged = list.concat(appends).sort(comparator)
  if(max && merged.size > max) {
    removes = merged.slice(max).concat(removes)
    merged = merged.take(max)
  }

  return {
    merged, appends, removes,
  }
}


