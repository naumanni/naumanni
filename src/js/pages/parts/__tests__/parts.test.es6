import React from 'react'
import renderer from 'react-test-renderer'

import {Account} from 'src/models'
import {IconFont, NowLoading, UserIconWithHost} from '../'


test('IconFont', () => {
  const tree = renderer.create(
    <IconFont iconName="iconA" />
  ).toJSON()
  expect(tree).toMatchSnapshot()

  const treeB = renderer.create(
    <IconFont iconName="iconB" />
  ).toJSON()
  expect(treeB).toMatchSnapshot()
})


test('NowLoading', () => {
  const tree = renderer.create(
    <NowLoading />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})


describe('UserIconWithHost', () => {
  it('can show user icon', () => {
    const mockAccount = new Account({
      acct: 'shn@oppai.tokyo',
      avatar: 'https://ot-mastodon.s3.amazonaws.com/accounts/avatars/000/000/001/original/2408e330e310f168.png?1492583237',
      display_name: 'Shin Adachi',
    })
    const tree = renderer.create(
      <UserIconWithHost account={mockAccount} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('can show default icon if not avatar was present', () => {
    const mockAccount = new Account({
      acct: 'shn@oppai.tokyo',
      avatar: null,
      display_name: 'Shin Adachi',
    })
    const tree = renderer.create(
      <UserIconWithHost account={mockAccount} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
