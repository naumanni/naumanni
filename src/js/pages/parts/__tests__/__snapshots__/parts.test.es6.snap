// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconFont 1`] = `
<span
  className="icon icon-iconA"
/>
`;

exports[`IconFont 2`] = `
<span
  className="icon icon-iconB"
/>
`;

exports[`NowLoading 1`] = `
<div
  className="niceLoader"
/>
`;

exports[`UserIconWithHost can show default icon if not avatar was present 1`] = `
<span
  className="userIcon with-host"
  onClick={undefined}
>
  <img
    alt="shn@oppai.tokyo"
    className="userIcon-avatar"
    src="/images/no-avatar.png"
    title="shn@oppai.tokyo"
  />
  <img
    className="userIcon-host"
    src="https://oppai.tokyo/favicon.ico"
  />
</span>
`;

exports[`UserIconWithHost can show user icon 1`] = `
<span
  className="userIcon with-host"
  onClick={undefined}
>
  <img
    alt="shn@oppai.tokyo"
    className="userIcon-avatar"
    src="https://ot-mastodon.s3.amazonaws.com/accounts/avatars/000/000/001/original/2408e330e310f168.png?1492583237"
    title="shn@oppai.tokyo"
  />
  <img
    className="userIcon-host"
    src="https://oppai.tokyo/favicon.ico"
  />
</span>
`;
